[#47] Add action to switch between Source/Header

ddscharfe opened the pull request:
- add server interface for clangd textDocument/switchSourceHeader extension
- add command and menu entry for switch action

Note: In the UI the term "toggle" is used instead of switch in order to be consistent with the CDT editor.

Fixes #47

jonahgraham commented:
I added a `Fixes #47` to the OP comment - having the `[#47]` in the subject isn't enough for GitHub to automatically link the PR to the Issue.

ddscharfe commented:
> I added a `Fixes #47` to the OP comment - having the `[#47]` in the subject isn't enough for GitHub to automatically link the PR to the Issue.

Thanks, @jonahgraham

ddscharfe commented:
Added a second commit to support cdt-lsp editors embedded in `MultiPageEditorPart` editors.

ghentschke commented:
LGTM. @ddscharfe  could you please stash the commits and provide a test for this stuff?

ruspl-afed commented on bundles/org.eclipse.cdt.lsp/META-INF/MANIFEST.MF:
```diff
@@ -4,7 +4,8 @@ Bundle-Name: Language Server based C/C++ Editor
 Bundle-SymbolicName: org.eclipse.cdt.lsp;singleton:=true
 Bundle-Version: 1.0.0.qualifier
 Export-Package: org.eclipse.cdt.lsp,
```
It could be reasonable to mark packages as `x-internal` until we have stable API

ddscharfe replied:
I think this might be a good idea, however I won't change this in this PR, as I think this should be a separate discussion.

ruspl-afed replied:
Another PR for `x-internal` is fine with me

ruspl-afed commented on bundles/org.eclipse.cdt.lsp/src/org/eclipse/cdt/lsp/services/ClangdLanguageServer.java:
```diff
@@ -0,0 +1,24 @@
+import org.eclipse.lsp4j.jsonrpc.services.JsonRequest;
+import org.eclipse.lsp4j.services.LanguageServer;
+
+public interface ClangdLanguageServer extends LanguageServer {
```
This interface pretends to be a part of public API. Do you plan to add any javadoc to describe it?

ddscharfe replied:
Good point. Not sure if it should be API, but it should get a javadoc nevertheless.

ddscharfe commented:
> LGTM. @ddscharfe could you please stash the commits and provide a test for this stuff?

I think you meant squash the commits - did that.

I added tests for the property tester. I did not write tests for the command handler, because it consists mostly of calling (static) methods. If we want to test code like this with unit tests we should use DI (which I think would be a good thing). What I'm not sure about if in the case of the handler it would make sense to create integration tests.
I introduced a dependency to mockito, I hope this is okay. I also had to add dependencies to `net.bytebuddy.byte-buddy(-agent)` to make mockito work. What I couldn't use is the Mockito Extension, because [mockito-junit-jupiter](https://search.maven.org/search?q=a:mockito-junit-jupiter) doesn't seem to be in the cdt target platform. Don't know what @jonahgraham thinks about this, maybe we can add it there?

I also added some more JavaDoc as suggested by @ruspl-afed.

ddscharfe commented:
Seems like the mockito dependency cannot be resolved, I was assuming we are using the cdt target platform?

ghentschke commented:
> Seems like the mockito dependency cannot be resolved, I was assuming we are using the cdt target platform?

I think this has to be changed since we are now part of cdt.

ruspl-afed commented:
@ddscharfe what if you add the required bundles to the target [here](https://github.com/eclipse-cdt/cdt-lsp/blob/master/releng/org.eclipse.cdt.lsp.target/org.eclipse.cdt.lsp.target.target) ? Then we can add them to the CDT target.

ddscharfe commented:
> @ddscharfe what if you add the required bundles to the target [here](https://github.com/eclipse-cdt/cdt-lsp/blob/master/releng/org.eclipse.cdt.lsp.target/org.eclipse.cdt.lsp.target.target) ? Then we can add them to the CDT target.

Thanks, I overlooked that.

I added the latest orbit repository for mockito, because the orbit repository (which is used for Guava and SnakeYAML) doesn't contain mockito-junit-jupiter.

ddscharfe commented:
Maven fails with this error:
```
java.lang.AbstractMethodError: Receiver class org.junit.jupiter.engine.descriptor.MethodExtensionContext does not define or inherit an implementation of the resolved method 'abstract java.util.Optional getTestInstances()' of interface org.junit.jupiter.api.extension.ExtensionContext.
```
I have no idea what this means, probably not using MockitoExtension would solve this problem. Any ideas?

ruspl-afed reviewed (approved):
LGTM

ruspl-afed commented:
Perhaps it was the mix of mockito versions. 
Should we merge it or do you plan more work here @ddscharfe ?

ddscharfe commented:
> Maven fails with this error:
> 
> ```
> java.lang.AbstractMethodError: Receiver class org.junit.jupiter.engine.descriptor.MethodExtensionContext does not define or inherit an implementation of the resolved method 'abstract java.util.Optional getTestInstances()' of interface org.junit.jupiter.api.extension.ExtensionContext.
> ```
> 
> I have no idea what this means, probably not using MockitoExtension would solve this problem. Any ideas?

I removed the MockitoExtension and use the `R20220531185310` orbit repository for the mockito dependency and now the build succeedes. If anyone knows how to solve the maven error, I'd be happy to hear any suggestions, because I like using @Mock. Otherwise we can leave it as it is. 
I didn't squash the commits to make this explicit in the history, hope this is okay.

ddscharfe commented:
Forgot to implement `LspEditorActiveTesterTest::editorPartWithInnerMatchingEditor` properly, fixed it.

jonahgraham commented:
> Don't know what @jonahgraham thinks about this, maybe we can add it there?

No problem adding the dependencies - with #51 (thanks @ruspl-afed) it will be easy to check that such new dependencies work well.

ghentschke commented:
@jonahgraham I think this PR can be rebased and merged. Could you do that please.

jonahgraham commented:
> @jonahgraham I think this PR can be rebased and merged. Could you do that please.

on it

jonahgraham commented:
> I didn't squash the commits to make this explicit in the history, hope this is okay.

I missed that on what I just force pushed - I am doing another force push now
